File: 2015/day9.rb
input = File.read('./input/day9.txt')

test_input = 'London to Dublin = 464
London to Belfast = 518
Dublin to Belfast = 141'

class Route
  def initialize(input)
    @paths = paths_from input
    @cities = @paths.map{|p| p[:cities] }.flatten.uniq
  end

  def paths_from(input)
    input.split("\n").map do |line|
      cities, distance = line.split(' = ')
      cities = cities.split(' to ')
      { cities: cities, distance: distance.to_i }
    end
  end

  def routes
    @cities.permutation
      .map{ |list| order_paths(list) }
      .reject{ |route| too_short(route) }.map do |route|
        route.reduce(0){|total, path| total + path[:distance] }
      end
  end

  def max
    routes.max
  end

  def min
    routes.min
  end

  def order_paths(list)
    list.each_with_index.map do |city, i|
      @paths.find{|path| path[:cities].include?(city) && path[:cities].include?(list[i - 1]) } unless i == 0
    end.compact
  end

  def too_short(route)
    route.length < @cities.length - 1
  end
end

test = Route.new(test_input)
puts test.min
actual = Route.new(input)
puts actual.min
puts test.max
puts actual.max

File: 2015/day9.js
const fs = require('fs')
const distances = fs.readFileSync('./input/day9.txt', 'utf8').trim().split("\n")

let options = {}
distances.forEach(distance => {
  const [from, _, to, __, km] = distance.split(' ')
  options[from] = options[from] || {}
  options[to] = options[to] || {}
  options[from][to] = Number(km)
  options[to][from] = Number(km)
})
const cities = Object.keys(options)

const routesStartingWith = cityArray => {
  const remainingCities = cities.filter(city => !cityArray.includes(city))
  if (remainingCities.length === 1) {
    return cityArray.concat(remainingCities)
  } else {
    return remainingCities.map(city => routesStartingWith(cityArray.concat(city)))
  }
}

const flatten = array => {
  if (Array.isArray(array[0][0])) {
    return array.reduce((total, elm) => total.concat(flatten(elm)), [])
  } else {
    return array
  }
}
const routes = flatten(cities.map(city => routesStartingWith([city])))

const routeLengths = routes.map(route => (
  route.reduce((distance, city, i) => (
    i === 0 ? distance : distance + options[route[i - 1]][city]
  ), 0)
))
const shortest = routeLengths.reduce((acc, curr) => Math.min(acc, curr))
const longest = routeLengths.reduce((acc, curr) => Math.max(acc, curr))
console.log(shortest, longest)
